### PR TITLE
Refactoring - remote lz composition

### PR DESCRIPTION
**Pull request by arnaudlh**
General internal refactoring to prioritize remote lz composition over local. 
As reported by @matze-li and @arne21a

### CODE REVIEW

**Comment by Matze-Li** [comment]
Hi Arnaud, 

Also the **Storage Account Network Rules** would need the remote lz: 

[/modules/storage_account/storage_account.tf](https://github.com/aztfmod/terraform-azurerm-caf/blob/23c0a8ba5f1cef5cc71c7a5b0d2bb54851a0744e/modules/storage_account/storage_account.tf#L162)

File: modules/analytics/databricks_workspace/workspace.tf [hunk 1]
```diff
@@ -48,48 +48,15 @@ resource "azurerm_databricks_workspace" "ws" {
...
-      )
-
-      public_subnet_name = try(coalesce(
-        try(var.settings.custom_parameters.public_subnet_name, null),
-        try(var.vnets[var.client_config.landingzone_key][var.settings.custom_parameters.vnet_key].subnets[var.settings.custom_parameters.public_subnet_key].name, null),
-        try(var.vnets[var.settings.custom_parameters.lz_key][var.settings.custom_parameters.vnet_key].subnets[var.settings.custom_parameters.public_subnet_key].name, null),
-        ),
-        null
-      )
+      virtual_network_id = can(var.settings.custom_parameters.virtual_network_id) || can(var.settings.custom_parameters.vnet_key) == false || can(var.settings.custom_parameters.vnet.key) == false ? try(var.settings.custom_parameters.virtual_network_id, null) : var.vnets[try(var.settings.custom_parameters.lz_key, var.client_config.landingzone_key)][var.settings.custom_parameters.vnet_key].id
```
Comment (LaurentLesle) [inline modules/analytics/databricks_workspace/workspace.tf:51]:
```suggestion
      virtual_network_id = can(var.settings.custom_parameters.virtual_network_id) || can(var.settings.custom_parameters.vnet_key) == false ? try(var.settings.custom_parameters.virtual_network_id, null) : var.vnets[try(var.settings.custom_parameters.lz_key, var.client_config.landingzone_key)][var.settings.custom_parameters.vnet_key].id
```

Was not in the original code and not in the refactored version. So I removed it
```
|| can(var.settings.custom_parameters.vnet.key) == false
```

File: modules/analytics/databricks_workspace/workspace.tf [hunk 2]
```diff
@@ -48,48 +48,15 @@ resource "azurerm_databricks_workspace" "ws" {
...
+      public_subnet_name = can(var.settings.custom_parameters.public_subnet_name) || can(var.settings.custom_parameters.vnet_key) == false || can(var.settings.custom_parameters.vnet.key) == false ? try(var.settings.custom_parameters.public_subnet_name, null) : var.vnets[try(var.settings.custom_parameters.lz_key, var.client_config.landingzone_key)][var.settings.custom_parameters.vnet_key].subnets[var.settings.custom_parameters.public_subnet_key].name
```
Comment (LaurentLesle) [inline modules/analytics/databricks_workspace/workspace.tf:52]:
```suggestion
      public_subnet_name = can(var.settings.custom_parameters.public_subnet_name) || can(var.settings.custom_parameters.vnet_key) == false ? try(var.settings.custom_parameters.public_subnet_name, null) : var.vnets[try(var.settings.custom_parameters.lz_key, var.client_config.landingzone_key)][var.settings.custom_parameters.vnet_key].subnets[var.settings.custom_parameters.public_subnet_key].name
```

Same as previous. Removed
```
|| can(var.settings.custom_parameters.vnet.key) == false
```

File: modules/analytics/databricks_workspace/workspace.tf [hunk 3]
```diff
@@ -48,48 +48,15 @@ resource "azurerm_databricks_workspace" "ws" {
...
-      )
-
-      private_subnet_name = try(coalesce(
-        try(var.settings.custom_parameters.private_subnet_name, null),
-        try(var.vnets[var.client_config.landingzone_key][var.settings.custom_parameters.vnet_key].subnets[var.settings.custom_parameters.private_subnet_key].name, null),
-        try(var.vnets[var.settings.custom_parameters.lz_key][var.settings.custom_parameters.vnet_key].subnets[var.settings.custom_parameters.private_subnet_key].name, null),
-        ),
-        null
-      )
+      public_subnet_network_security_group_association_id = can(var.settings.custom_parameters.public_subnet_network_security_group_association_id) || can(var.settings.custom_parameters.public_subnet_key) == false || can(var.settings.custom_parameters.public_subnet.key) == false ? try(var.settings.custom_parameters.public_subnet_network_security_group_association_id, null) : var.vnets[try(var.settings.custom_parameters.lz_key, var.client_config.landingzone_key)][var.settings.custom_parameters.vnet_key].subnets[var.settings.custom_parameters.public_subnet_key].id
```
Comment (LaurentLesle) [inline modules/analytics/databricks_workspace/workspace.tf:55]:
```suggestion
      public_subnet_network_security_group_association_id = can(var.settings.custom_parameters.public_subnet_network_security_group_association_id) || can(var.settings.custom_parameters.public_subnet_key) == false ? try(var.settings.custom_parameters.public_subnet_network_security_group_association_id, null) : var.vnets[try(var.settings.custom_parameters.lz_key, var.client_config.landingzone_key)][var.settings.custom_parameters.vnet_key].subnets[var.settings.custom_parameters.public_subnet_key].id
```

Same a previous. Removed
```
 || can(var.settings.custom_parameters.public_subnet.key) == false 
```

File: modules/analytics/databricks_workspace/workspace.tf [hunk 4]
```diff
@@ -48,48 +48,15 @@ resource "azurerm_databricks_workspace" "ws" {
...
+      private_subnet_name                                 = can(var.settings.custom_parameters.private_subnet_name) || can(var.settings.custom_parameters.private_subnet_key) == false || can(var.settings.custom_parameters.private_subnet.key) == false ? try(var.settings.custom_parameters.private_subnet_name, null) : var.vnets[try(var.settings.custom_parameters.lz_key, var.client_config.landingzone_key)][var.settings.custom_parameters.vnet_key].subnets[var.settings.custom_parameters.private_subnet_key].name
```
Comment (LaurentLesle) [inline modules/analytics/databricks_workspace/workspace.tf:56]:
```suggestion
      private_subnet_name                                 = can(var.settings.custom_parameters.private_subnet_name) || can(var.settings.custom_parameters.private_subnet_key) == false ? try(var.settings.custom_parameters.private_subnet_name, null) : var.vnets[try(var.settings.custom_parameters.lz_key, var.client_config.landingzone_key)][var.settings.custom_parameters.vnet_key].subnets[var.settings.custom_parameters.private_subnet_key].name
```

Removed 
```
 || can(var.settings.custom_parameters.private_subnet.key) == false 
```

File: modules/analytics/databricks_workspace/workspace.tf [hunk 5]
```diff
@@ -48,48 +48,15 @@ resource "azurerm_databricks_workspace" "ws" {
...
-      private_subnet_network_security_group_association_id = try(coalesce(
-        try(var.settings.custom_parameters.private_subnet_network_security_group_association_id, null),
-        try(var.vnets[var.client_config.landingzone_key][var.settings.custom_parameters.vnet_key].subnets[var.settings.custom_parameters.public_subnet_key].id, null),
-        try(var.vnets[var.settings.custom_parameters.lz_key][var.settings.custom_parameters.vnet_key].subnets[var.settings.custom_parameters.public_subnet_key].id, null),
-        ),
-        null
-      )
-
+      private_subnet_network_security_group_association_id = can(var.settings.custom_parameters.private_subnet_network_security_group_association_id) || can(var.settings.custom_parameters.public_subnet_key) == false || can(var.settings.custom_parameters.public_subnet.key) == false ? try(var.settings.custom_parameters.private_subnet_network_security_group_association_id, null) : var.vnets[try(var.settings.custom_parameters.lz_key, var.client_config.landingzone_key)][var.settings.custom_parameters.vnet_key].subnets[var.settings.custom_parameters.public_subnet_key].id
```
Comment (LaurentLesle) [inline modules/analytics/databricks_workspace/workspace.tf:59]:
```suggestion
      private_subnet_network_security_group_association_id = can(var.settings.custom_parameters.private_subnet_network_security_group_association_id) || can(var.settings.custom_parameters.public_subnet_key) == false ? try(var.settings.custom_parameters.private_subnet_network_security_group_association_id, null) : var.vnets[try(var.settings.custom_parameters.lz_key, var.client_config.landingzone_key)][var.settings.custom_parameters.vnet_key].subnets[var.settings.custom_parameters.public_subnet_key].id
```

Two problems here:

1. Removed this as not in the original and target implementation

```
|| can(var.settings.custom_parameters.public_subnet.key) == false
```

2. mixing private and public
private_subnet_network_security_group_association_id and using
var.settings.custom_parameters.public_subnet_key
Should it be var.settings.custom_parameters.private_subnet_key

File: modules/compute/virtual_machine_scale_set/vmss_windows.tf [hunk 6]
```diff
@@ -135,8 +135,7 @@ resource "azurerm_windows_virtual_machine_scale_set" "vmss" {
   }
 
   source_image_id = try(each.value.source_image_reference, null) == null ? format("%s%s",
-    try(each.value.custom_image_id, var.image_definitions[var.client_config.landingzone_key][each.value.custom_image_key].id,
-    var.image_definitions[each.value.custom_image_lz_key][each.value.custom_image_key].id),
+    try(each.value.custom_image_id, var.image_definitions[try(each.value.custom_image_lz_ke, var.client_config.landingzone_key)][each.value.custom_image_key].id),
```
Comment (LaurentLesle) [inline modules/compute/virtual_machine_scale_set/vmss_windows.tf:138]:
```suggestion
    try(each.value.custom_image_id, var.image_definitions[try(each.value.custom_image_lz_key, var.client_config.landingzone_key)][each.value.custom_image_key].id),
```

File: modules/databases/data_explorer/kusto_clusters/module.tf [hunk 7]
```diff
@@ -38,7 +38,7 @@ resource "azurerm_kusto_cluster" "kusto" {
   dynamic "virtual_network_configuration" {
     for_each = try(var.settings.virtual_network_configuration, null) != null ? [var.settings.virtual_network_configuration] : []
     content {
-      subnet_id                    = try(virtual_network_configuration.value.vnet_key, null) == null ? null : try(var.combined_resources.vnets[var.client_config.landingzone_key][virtual_network_configuration.value.vnet_key].subnets[virtual_network_configuration.value.subnet_key].id, var.combined_resources.vnets[virtual_network_configuration.value.lz_key][virtual_network_configuration.value.vnet_key].subnets[virtual_network_configuration.value.subnet_key].id)
+      subnet_id                    = try(virtual_network_configuration.value.vnet_key, null) == null ? null : try(var.combined_resources.vnets[try(virtual_network_configuration.value.lz_key, var.client_config.landingzone_key)][virtual_network_configuration.value.vnet_key].subnets[virtual_network_configuration.value.subnet_key].id, null)
```
Comment (LaurentLesle) [inline modules/databases/data_explorer/kusto_clusters/module.tf:41]:
```suggestion
      subnet_id                    = can(virtual_network_configuration.value.subnet_id) || can(virtual_network_configuration.value.subnet_key) == false ?  try(virtual_network_configuration.value.subnet_id, null) : try(virtual_network_configuration.value.vnet_key, null) == null ? null : var.combined_resources.vnets[try(virtual_network_configuration.value.lz_key, var.client_config.landingzone_key)][virtual_network_configuration.value.vnet_key].subnets[virtual_network_configuration.value.subnet_key].id
```

**Review by LaurentLesle — CHANGES_REQUESTED** [review]
Very good refactoring. Some comments

File: container_groups.tf [hunk 8]
```diff
@@ -3,13 +3,14 @@ module "container_groups" {
...
 
-  base_tags                = try(local.global_settings.inherit_tags, false) ? local.combined_objects_resource_groups[try(each.value.lz_key, local.client_config.landingzone_key)][each.value.resource_group_key].tags : {}
+
+  location                 = can(local.global_settings.regions[each.value.region]) ? local.global_settings.regions[each.value.region] : local.combined_objects_resource_groups[try(each.value.resource_group.lz_key, local.client_config.landingzone_key)][try(each.value.resource_group.key, each.value.resource_group_key)].location
+  resource_group_name      = can(each.value.resource_group.name) || can(each.value.resource_group_name) ? try(each.value.resource_group.name, each.value.resource_group_name) : local.combined_objects_resource_groups[try(each.value.resource_group.lz_key, local.client_config.landingzone_key)][try(each.value.resource_group_key, each.value.resource_group.key)].name
+  base_tags                = try(local.global_settings.inherit_tags, false) ? local.combined_objects_resource_groups[try(each.value.resource_group.lz_key, local.client_config.landingzone_key)][try(each.value.resource_group.key, each.value.resource_group_key)].tags : {}
```
[the remark this answers is not in the document]
Comment (arnaudlh) [reply container_groups.tf:9]: Good catch, will stash and investigate on another PR as it will need more testing.

**Review by LaurentLesle — APPROVED** [review]
LGTM

**Pull request merged**